packages/core/core/src/TargetResolver.js

```diff
@@ -37,6 +41,228 @@ const DEFAULT_PRODUCTION_ENGINES = {
 const DEFAULT_DIST_DIRNAME = 'dist';
 const COMMON_TARGETS = ['main', 'module', 'browser', 'types'];
 
+function parseEngines(
+  targetName: ?string,
+  engines: mixed
+): Engines | typeof undefined {
+  if (engines === undefined) {
+    return;
+  } else if (engines && typeof engines === 'object') {
+    let result = {};
+    for (let key in engines) {
+      let value: mixed = engines[key];
+      if (key === 'browsers' && Array.isArray(value)) {
+        value = value.map(v => {
+          if (typeof v !== 'string') {
+            throw new Error(
+              `Invalid value in engines.browsers array ${
+                targetName ? `for target "${targetName}"` : ''
+              }: ${String(v)}`
+            );
+          }
+          return v;
+        });
+      } else if (typeof value !== 'string') {
+        throw new Error(
+          `Invalid value for engines.${key} ${
+            targetName ? `for target "${targetName}"` : ''
+          }: ${String(value)}`
+        );
+      }
+      result[key] = value;
+    }
+    return result;
+  } else {
+    throw new Error(
+      `Invalid engines ${
+        targetName ? `for target "${targetName}"` : ''
+      }: ${String(engines)}`
+    );
+  }
+}
+
+function parseDescriptor(
+  targetName: string,
+  descriptor: mixed
+): {|
+  context?: EnvironmentContext | typeof undefined,
```

Comment:
So far we haven't been differentiating between null and undefined, using `?EnvironmentContext` in a case like this

Reply:
Because most properties in `TargetDescriptor` are optional and have a non-nullish type, for example:

This fails:
```js
type Opts = {
  context?: string
};

let value: ?string = null;

let opts: Opts = {
  context: value
}
```

This works:
```js
type Opts = {
  context?: string
};

let value: string | typeof undefined = ”xyz";

let opts: Opts = {
  context: value
}
```

Reply:
I think the `Opts` type needs to be this:

```js
type Opts = {
  context?: ?string
}
```